Service/app/models/aci/snapshots.py

```diff
@@ -63,6 +68,104 @@ def before_snapshot_create(data, **kwargs):
+            except Exception as err:
+                abort(400,str(err))
+    abort(400, "File metadata not found")
+
+def upload_snapshot():
+    from ..utils import get_app_config
+    if request.files is None:
+        abort(400, "No files uploaded")
+    else:
+        tmp_dir = os.path.realpath("%s/%s" % (current_app.config["TMP_DIR"],uuid.uuid4()))
+        config = get_app_config()
+        dst = config.get("DATA_DIR", "/tmp/")
+        try:
+            for filename in request.files:
+                f = request.files[filename] 
+                # validate file is .tgz file and allowed name
+                if not re.search("^[^\/]+\.tgz$", f.filename):
+                    abort(400, "Invalid filename for snapshot: %s" % f.filename)
+                temp_filename = os.path.join(tmp_dir , f.filename)
+                # create the directory
+                os.mkdir(tmp_dir)
+                os.chdir(tmp_dir)
+                # unzip and stuff..
+                f.save(temp_filename)
+                if aci_utils.run_command('tar --force-local -zxf ' + temp_filename) is None:
+                    logger.debug("Traceback:\n %s",traceback.format_exc())
+                    abort(500,"Failed to unzip snapshot")
+                if aci_utils.run_command('tar --force-local -zxf bundle.tgz') is None:
+                    logger.debug("Traceback: \n %s", traceback.format_exc())
+                    abort(500,"Failed to unzip snapshot")
+                # try to extract required data, abort if not in correct format which triggers exception
+                try:
+                    checksum = json.loads(open( os.path.join(tmp_dir,'md5checksum.json')).read()).get('md5checksum')
+                    bundle_checksum = aci_utils.get_file_md5('bundle.tgz') 
+                    if bundle_checksum != checksum:
+                        abort(400,"snapshot file md5 is invalid")
+                    fnew = open(os.path.join(tmp_dir,'snapshot.json'), 'r')
+                    fileJson = json.loads(fnew.read())
+                except BadRequest as e: raise e
+                except Exception as e:
+                    logger.debug("Traceback:\n %s", traceback.format_exc())
+                    abort(400, "Invalid snapshot file")
+
+                for node in fileJson.get('nodes'):
+                    if os.path.exists(os.path.join(tmp_dir,'node-') + node) is False:
+                        abort(400,'One or more nodes missing in provided snapshots')
+                    snap = Snapshots()
+                    supported_attr = ["fabric","definition", "description", "fabric_domain", "nodes",
+                                        "start_time", "wait_time", "total_time"]
+                    for attr in supported_attr:
+                        if attr in fileJson:
+                            setattr(snap, attr, fileJson[attr])
+                    # ensure fabric and definition exists as that will trigger save to fail
```

Comment:
I would send snap.to_json() here instead of fileJson.  there's no guarantee that fileJson has the fields present so it trigger a keyerror exception in the validation in before_snapshot_create which isn't the error we're going for here.